src/libsyntax/tokenstream.rs

```diff
@@ -820,98 +929,108 @@ impl Index<usize> for TokenStream {
     type Output = TokenTree;
 
     fn index(&self, index: usize) -> &TokenTree {
```

Comment:
Mainly index-based iteration instead of using the iterator interface for which `next` can be made constant-time (amortized).
On IRC we've also discussed a potential flattened `TokenTree` scheme, in which indices in the rope would not correspond to the top-level `TokenTree`'s elements, but flattened elements.